### DISCUSSION TITLE
How to hide "frames excluded" message from logs?

### DISCUSSION THREAD

**Discussion by hexwit** [logfellow/logstash-logback-encoder#862] [Q&A] [answered]
I am using ShortenedThrowableConverter with LoggingEventCompositeJsonEncoder to configure stacktrace frames logged. 
And I see messages like "... n frames excluded". 
Sometimes it is only 1 frame, and such messages bring almost no use in this case.

I'd like to know, is there any config that will prevent printing messages like this?

**Comment by brenuart** [comment]
The "_... n frames excluded_" message tells you how many frames have been excluded due to the configured `<excludes>` patterns - it is also an indication that the stacktrace is not complete and elements have been removed.

Could you please tell us a bit more why you found these messages annoying (maybe with an example)?

**Reply by hexwit** [reply]
The use case is very simple: I am removing frames that have no any sense to me, and in some cases it is only 1 appearance of some frame in a block of frames. So I see message "... 1 frames excluded" multiple times. It is basically the same if I would keep frame without exclusion. In such case the size of a stack trace is not reduced.

Lots of frames, like ^java\.lang\.Thread\.run.* has very little value for the debug, so they are safely can be removed. 
But such cases usually produce messages with 1 frame excluded. 

What could be done instead: 
 - configuration that will hide such messages
 - configuration that will add (customisable?) message at the end of stack trace indicating that it is not full, and some frames were dropped. So now developer knows that something missing + has short stack trace. 
 
Here is why it is ok: when you see a message "... 12 frames excluded" you actually don't know what they are. 
And position of excluded frames does not gives you much. As I see, it is not possible to guess what exactly has been removed, even having access to the configuration.

Please let me know if I need to provide more details about my case.

**Reply by hexwit** [reply]
From the another post I found out that at least two statements should match to be excluded from a stack trace. 
So the problem description a little inaccurate, but a suggestion about configuration still actual. 

If you check following stack trace, you'll see, that one single line repeats among others. (FilterChainProxy)
```
at org.springframework.security.web.access.intercept.FilterSecurityInterceptor.invoke(FilterSecurityInterceptor.java:115)
	at org.springframework.security.web.access.intercept.FilterSecurityInterceptor.doFilter(FilterSecurityInterceptor.java:81)
	at org.springframework.security.web.FilterChainProxy$VirtualFilterChain.doFilter(FilterChainProxy.java:346)
	at org.springframework.security.web.access.ExceptionTranslationFilter.doFilter(ExceptionTranslationFilter.java:122)
	at org.springframework.security.web.access.ExceptionTranslationFilter.doFilter(ExceptionTranslationFilter.java:116)
	at org.springframework.security.web.FilterChainProxy$VirtualFilterChain.doFilter(FilterChainProxy.java:346)
	at org.springframework.security.web.session.SessionManagementFilter.doFilter(SessionManagementFilter.java:126)
	at org.springframework.security.web.session.SessionManagementFilter.doFilter(SessionManagementFilter.java:81)
	at org.springframework.security.web.FilterChainProxy$VirtualFilterChain.doFilter(FilterChainProxy.java:346)
	at org.springframework.security.web.authentication.AnonymousAuthenticationFilter.doFilter(AnonymousAuthenticationFilter.java:109)
	at org.springframework.security.web.FilterChainProxy$VirtualFilterChain.doFilter(FilterChainProxy.java:346)
```
It would be nice to get rid of FilterChainProxy lines, as it useless here. But seems to be I can't do that. 

Could you please tell, is there any chance to have a solution for removing single lines and ignoring "... n frames excluded" message?

**Reply by hexwit** [reply]
@brenuart could you please review my comments?

**Reply by brenuart** [reply]
Hi @hexwit, unfortunately there is currently no way to achieve what you are trying to do.
The "exclude" feature is more a way to "collapse" multiple consecutive lines and replace them with a shorter message.

From what I understand, you're looking for a way to filter out certain items from the stack trace and pretend they didn't even exist. This process would occur before the stacktrace is rendered and additional groups of lines possibly collapsed (the _exclude_ feature).
This seems legit to me, especially to remove traces originating from generated classes like `$$SpringCGLIB$$`. In these cases, in doesn't make sense to replace the filtered lines by a message as does the "exclude" feature.

I'm still thinking about it but maybe we should add a `<filter>` feature alongside the current `<exclude>` (names may be confusing tho - tbd)...

In your original message you say:
> So I see message "... 1 frames excluded" multiple times.

This is rather strange knowing that the converter waits for at least two consecutive excluded lines to collapse them. Do you have an example of when a single line is excluded? This might be a bug...

**Reply by hexwit** [reply]
@brenuart Thank you very much for so detailed answer.

> This is rather strange knowing that the converter waits for at least two consecutive excluded lines to collapse them. Do you have an example of when a single line is excluded? This might be a bug...

I did so many different tests so I can't provide any meaningful right details right now. It could be that it was my mistake. If I will face this situation again, I will file an issue. 


You are right, I was looking a way, how to remove autogenerated classes, and frames that has no meaning in the debugging process. 
The name of 'exclude' tag made me think that it going to removes lines from the stack trace. Maybe 'collapse' would have more sense for its functionality. 

Could you please tell, is there any chance to expect <filter> tag (or something for removing lines from stacktrace) in the nearest future?

**Comment by hexwit** [comment, marked as answer]
As no existing solution available, and not planned, I end up with implementing own solution with filtering possibilities. 
It is based on ThrowableHandlingConverter. I used ShortenedThrowableConverter as a reference.
